### PR TITLE
Add support for dotnet user secrets files

### LINKED ISSUE #1970: Add support for user secret file

### PR DESCRIPTION
**Pull request by gzuber**
Resolves #1970

This will allow a user to make use of dotnet user secrets files detailed [here](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-3.1&tabs=windows). The user would also have to implement dependency injection in their function ([more info](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection)), and access their development secrets through an `IConfiguration` object.

One thing I'm unsure of is the logging I added. If that's the right way to do it, but also if those logs are clear and helpful. I want to give the user insight into the decision making process/where we looked so they are empowered to fix any issues, but I didn't want to bombard them with a bunch of "checking this folder... checking this folder..."

### CODE REVIEW

**Comment by ahmelsayed** [comment]
Just to confirm, this is only for dotnet apps, right?

File: src/Azure.Functions.Cli/Secrets/UserSecretsConfigurationBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Azure.Functions.Cli.Diagnostics
+{
+    internal class UserSecretsConfigurationBuilder : IConfigureBuilder<IConfigurationBuilder>
+    {
+        private readonly string _scriptPath;
+
+        public UserSecretsConfigurationBuilder(string scriptPath)
+        {
+            _scriptPath = scriptPath;
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Secrets/UserSecretsConfigurationBuilder.cs:15]: nit: null validation

File: src/Azure.Functions.Cli/Secrets/UserSecretsConfigurationBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            string userSecretsId = GetUserSecretsId(_scriptPath);
+            if (userSecretsId == null) return;
+
+            builder.AddUserSecrets(userSecretsId);
+        }
+
+        private string GetUserSecretsId(string scriptPath)
+        {
+            string projectFilePath = ProjectHelpers.FindProjectFile(scriptPath);
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Secrets/UserSecretsConfigurationBuilder.cs:28]: null validation

File: src/Azure.Functions.Cli/Helpers/ProjectHelpers.cs [hunk 3]
```diff
@@ -4,16 +4,37 @@
...
-            ProjectItemElement existingPackageReference = project.Items
-                .FirstOrDefault(item => item.ItemType == Constants.PackageReferenceElementName && item.Include.ToLowerInvariant() == packageId.ToLowerInvariant());
-            return existingPackageReference != null;
+            ColoredConsoleLogger logger = new ColoredConsoleLogger("ProjectHelpers");
+            DirectoryInfo filePath = new DirectoryInfo(path);
+
+            do
+            {
+                var projectFiles = filePath.GetFiles("*.csproj");
+                if (projectFiles.Count() > 0)
```
Comment (fabiocav) [inline src/Azure.Functions.Cli/Helpers/ProjectHelpers.cs:23]: nit: `projectFiles.Any()`

**Comment by gzuber** [comment]
> Just to confirm, this is only meaningful for dotnet apps, right?

Correct

**Comment by soninaren** [comment]
@gzuber can you please cherry pick changes to v2 - dev branch as well?

**Comment by vijayrkn** [comment]
Tested with the latest build and it read the conn string from the user secrets but I still get this warning. Is it something that can be suppressed if the value is present in the user secrets?

`Warning: Cannot find value named 'ConnectionStrings:MyQueueConn' in local.settings.json that matches 'connection' property set on 'queueTrigger' in 'C:\Users\userid\source\repos\FunctionApp46\FunctionApp46\bin\Debug\netcoreapp3.1\Function\function.json'. You can run 'func azure functionapp fetch-app-settings <functionAppName>' or specify a connection string in local.settings.json.`

**Comment by gzuber** [comment]
@vijayrkn Any chance you can share a quick repo? I haven't run into that one yet in my own testing, would love to clear that up

**Comment by vijayrkn** [comment]
sure. pretty standard setup. The only thing different I can think of is the nested config:

Function: 
```
    public static class Function
    {
        [FunctionName("Function")]
        public static void Run([QueueTrigger("myqueue-items", Connection = "ConnectionStrings:MyQueueConn")]string myQueueItem, ILogger log)
        {
            log.LogInformation($"C# Queue trigger function processed: {myQueueItem}");
        }
    }
```

local.settings.json

```
{
    "IsEncrypted": false,
    "Values": {
        "AzureWebJobsStorage": "UseDevelopmentStorage=true",
        "FUNCTIONS_WORKER_RUNTIME": "dotnet"
    }
}
```

user secrets: 
```
{
  "ConnectionStrings:MyQueueConn": "UseDevelopmentStorage=true"
}
```

csproj: 
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
    <UserSecretsId>4a9f61f7-0f22-4d2d-bc40-6ef82a58df03</UserSecretsId>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="3.0.10" />
    <PackageReference Include="Microsoft.Extensions.Configuration.UserSecrets" Version="3.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.7" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>

```

Let me know if you need anything more.

**Pull request merged**